### PR TITLE
MSTR-11: Add Billing Contact Address Fields

### PR DESCRIPTION
**Pull request by ConR24**

### CODE REVIEW

File: src/apps/myaccount/i18n/fr-FR.json [hunk 1]
```diff
@@ -82,7 +87,12 @@
 
 		"accountAdministrator": "Changer carte",
 		"accountAdministratorEmail": "Email",
-		"accountAdministratorNumber": "Numéro de téléphone"
+		"accountAdministratorNumber": "Numéro de téléphone",
+		"accountAdministratorAddress":"Adresse",
+		"accountAdministratorAddressLineTwo":"Ligne d'adresse 2",
+		"accountAdministratorAddressCity":"Ville",
+		"accountAdministratorAddressState":"Etat",
```
Comment (azefiel) [inline src/apps/myaccount/i18n/fr-FR.json]:
```suggestion
		"accountAdministratorAddressState":"Région",
```

File: src/apps/myaccount/i18n/fr-FR.json [hunk 2]
```diff
@@ -82,7 +87,12 @@
...
+		"accountAdministratorAddressZipCode":"code postal"
```
Comment (azefiel) [inline src/apps/myaccount/i18n/fr-FR.json]:
```suggestion
		"accountAdministratorAddressZipCode":"Code postal"
```

File: src/apps/myaccount/submodules/billing/views/layout.html [hunk 3]
```diff
@@ -102,6 +102,36 @@
 									<input type="text" id="account_administrator_number" name="contact.billing.number" value="{{account.contact.billing.number}}" placeholder="{{ i18n.placeholderNumber }}"/>
 								</div>
 							</div>
+							<div class="control-group">
+								<label class="control-label" for="account_administrator_address">{{ i18n.billing.accountAdministratorAddress }}</label>
+								<div class="controls">
+									<input type="text" id="account_administrator_address" name="contact.billing.address" value="{{account.contact.billing.address}}" placeholder="{{ i18n.placeholderAddress }}" />
```
Comment (azefiel) [inline src/apps/myaccount/submodules/billing/views/layout.html]:
```suggestion
									<input type="text" id="account_administrator_address" name="contact.billing.street_address" value="{{account.contact.billing.street_address}}" placeholder="{{ i18n.placeholderAddress }}" />
```

File: src/apps/myaccount/submodules/billing/views/layout.html [hunk 4]
```diff
@@ -102,6 +102,36 @@
...
+								</div>
+							</div>
+							<div class="control-group">
+								<label class="control-label" for="account_administrator_address_line_2">{{ i18n.billing.accountAdministratorAddressLineTwo  }}</label>
+								<div class="controls">
+									<input type="text" id="account_administrator_address_line_2" name="contact.billing.adddressLineTwo" value="{{account.contact.billing.addressLineTwo}}" placeholder="{{ i18n.placeholderAddressLineTwo  }}"/>
```
Comment (azefiel) [inline src/apps/myaccount/submodules/billing/views/layout.html]:
```suggestion
									<input type="text" id="account_administrator_address_line_2" name="contact.billing.street_address_extra" value="{{account.contact.billing.street_address_extra}}" placeholder="{{ i18n.placeholderAddressLineTwo  }}"/>
```

File: src/apps/myaccount/submodules/billing/views/layout.html [hunk 5]
```diff
@@ -102,6 +102,36 @@
...
+								</div>
+							</div>
+							<div class="control-group">
+								<label class="control-label" for="account_administrator_city">{{ i18n.billing.accountAdministratorAddressCity  }}</label>
+								<div class="controls">
+									<input type="text" id="account_administrator_city" name="contact.billing.city" value="{{acocunt.contact.billing.city}}" placeholder="{{ i18n.placeholderCity  }}" />
```
Comment (azefiel) [inline src/apps/myaccount/submodules/billing/views/layout.html]:
```suggestion
									<input type="text" id="account_administrator_city" name="contact.billing.locality" value="{{acocunt.contact.billing.locality}}" placeholder="{{ i18n.placeholderCity  }}" />
```

File: src/apps/myaccount/submodules/billing/views/layout.html [hunk 6]
```diff
@@ -102,6 +102,36 @@
...
+								</div>
+							</div>
+							<div class="control-group">
+								<label class="control-label" for="account_administrator_state">{{ i18n.billing.accountAdministratorAddressState  }}</label>
+								<div class="controls">
+									<input type="text" id="account_administrator_state" name="contact.billing.state" value="{{account.contact.billing.state}}" placeholder="{{ i18n.placeholderState  }}"/>
```
Comment (azefiel) [inline src/apps/myaccount/submodules/billing/views/layout.html]:
```suggestion
									<input type="text" id="account_administrator_state" name="contact.billing.region" value="{{account.contact.billing.region}}" placeholder="{{ i18n.placeholderState  }}"/>
```

File: src/apps/myaccount/submodules/billing/views/layout.html [hunk 7]
```diff
@@ -102,6 +102,36 @@
...
+								</div>
+							</div>
+							<div class="control-group">
+								<label class="control-label" for="account_administrator_zipcode">{{ i18n.billing.accountAdministratorAddressZipCode }}</label>
+								<div class="controls">
+									<input type="text" id="account_administrator_zipcode" name="contact.billing.zipcode" value="{{account.contact.billing.zipcode}}" placeholder="{{ i18n.placeholderZipCode }}"/>
```
Comment (azefiel) [inline src/apps/myaccount/submodules/billing/views/layout.html]:
```suggestion
									<input type="text" id="account_administrator_zipcode" name="contact.billing.postal_code" value="{{account.contact.billing.postal_code}}" placeholder="{{ i18n.placeholderZipCode }}"/>
```

File: src/apps/myaccount/i18n/en-US.json [hunk 8]
```diff
@@ -168,7 +173,12 @@
 
 		"accountAdministrator": "Billing Contact",
 		"accountAdministratorEmail": "Billing Contact Email",
-		"accountAdministratorNumber": "Phone Number"
+		"accountAdministratorNumber": "Phone Number",
+		"accountAdministratorAddress": "Address",
```
Comment (azefiel) [inline src/apps/myaccount/i18n/en-US.json]:
```suggestion
		"accountAdministratorAddress": "Primary Service Address",
```

File: src/apps/myaccount/i18n/en-US.json [hunk 9]
```diff
@@ -168,7 +173,12 @@
...
+		"accountAdministratorAddressLineTwo": "Address Line 2",
+		"accountAdministratorAddressCity": "City",
+		"accountAdministratorAddressState": "State",
```
Comment (azefiel) [inline src/apps/myaccount/i18n/en-US.json]:
```suggestion
		"accountAdministratorAddressState": "State/Region/Province",
```

File: src/apps/myaccount/i18n/en-US.json [hunk 10]
```diff
@@ -168,7 +173,12 @@
...
+		"accountAdministratorAddressZipCode": "Zip Code"
```
Comment (azefiel) [inline src/apps/myaccount/i18n/en-US.json]:
```suggestion
		"accountAdministratorAddressZipCode": "Zip/Postal Code"
```

**Comment by azefiel** [comment]
While you are at it, you can add the following fields as well since we are going to make them mandatory with our new account creation wizard:

- `account.contact.billing.name`: "Full Name"
- `account.contact.billing.country`: "Country"

And you can simplify the current email filed label from "Billing Contact Email" to "Email"

**Comment by azefiel** [comment]
Once you are done with the changes requested in my review, you'll have to set up validation (basically, every billing related field is required, ZIP code should only contain digits and email format should be enforced).

To do so, you'll have to use `monster.ui#validate` to define validation rules for your form ([doc](https://docs.2600hz.com/ui/docs/validation/)).

All of this will have to take place in the main myaccount `app.js` because of how the current system was implemented:
https://github.com/2600hz/monster-ui/blob/master/src/apps/myaccount/app.js#L805

Let me know if you have question!

**Comment by ConR24** [comment]
I defined the rules for the billing form in its own function similar to the password validate function and I added an elseif statement to the function you linked to, but it seems that it is ignoring the rules that I set and I'm not sure why.

**Comment by azefiel** [comment]
Can you push it to the feature branch so it is visible on here?

File: src/apps/myaccount/app.js [hunk 11]
```diff
@@ -786,6 +786,45 @@ define(function(require) {
 			}
 		},
 
+		validateAccountAdministratorForm(formAccountAdministrator, callback) {
```
Comment (ConR24) [inline src/apps/myaccount/app.js]: Here is the function where I define the validation rules

File: src/apps/myaccount/app.js [hunk 12]
```diff
@@ -786,6 +786,45 @@ define(function(require) {
...
+					    required: true
+					},
+					'contact.billing.postal_code': {
+					    required: true,
+					    digits: true
+					}
+				}
+			});
+			console.log(monster.ui.validate(formAccountAdministrator));
+			if (monster.ui.validate(formAccountAdministrator)) {
```
Comment (azefiel) [inline src/apps/myaccount/app.js]: Here need to use `monster.ui#valid` to trigger the form validation process.

File: src/apps/myaccount/app.js (hunk 11 above)
Comment (ConR24) [inline src/apps/myaccount/app.js]: Here is where I define the validation rules for the form

File: src/apps/myaccount/app.js [hunk 13]
```diff
@@ -803,13 +842,18 @@ define(function(require) {
...
-
+					var formAccountAdministrator = template.find('#form_account_administrator');
+					
 					// This is still ghetto, I didn't want to re-factor the whole code to tweak the validation
 					// If the field is password, we start custom validation
+					
 					if (formPassword.length) {
 						self.validatePasswordForm(formPassword, callback);
 					// otherwise we don't have any validation for this field, we execute the callback
+					} else if (formAccountAdministrator.length) {
```
Comment (ConR24) [inline src/apps/myaccount/app.js]: Here is where I call my validate function

**Comment by ConR24** [comment]
Just made that change, everything is working now

**Comment by ConR24** [comment]
I pushed up the validation fixes if you want to take a look.

**Comment by azefiel** [comment]
Seems like the minification task is throwing an error...

File: src/apps/myaccount/app.js (hunk 11 above)
Comment (azefiel) [inline src/apps/myaccount/app.js]: Please don't use object method shorthand as our build does not include a transpiling step.
```suggestion
		validateAccountAdministratorForm: function(formAccountAdministrator, callback) {
```

Comment (ConR24) [reply src/apps/myaccount/app.js]: Just pushed up a fix for this

**Comment by ConR24** [comment]
Checks have passed

**Comment by azefiel** [comment]
Good to go!

**Comment by ConR24** [comment]
Awesome, I should do a squash and merge right?

**Comment by azefiel** [comment]
Yes, it shouldn't give you any other option actually.

**Comment by ConR24** [comment]
Ok, Thank you so much for all the help!

**Pull request merged**